File: scripts/keepalive.py
#!/usr/bin/env python3
"""Keep-alive pings for Supabase Storage and MongoDB."""
import sys
import json
import os
from pathlib import Path
from urllib.request import Request, urlopen
from urllib.error import HTTPError

SCRIPT_DIR = Path(__file__).resolve().parent


def keep_alive_supabase():
    url = os.environ["SUPABASE_URL"]
    key = os.environ["SUPABASE_SERVICE_ROLE_KEY"]
    bucket = "pdfs"
    storage_path = "keepalive/sample.pdf"
    pdf_bytes = (SCRIPT_DIR / "sample.pdf").read_bytes()

    # Upload
    upload_url = f"{url}/storage/v1/object/{bucket}/{storage_path}"
    req = Request(upload_url, data=pdf_bytes, method="POST")
    req.add_header("Authorization", f"Bearer {key}")
    req.add_header("Content-Type", "application/pdf")
    req.add_header("x-upsert", "true")
    try:
        resp = urlopen(req)
        print("[supabase] Upload OK:", resp.read().decode())
    except HTTPError as e:
        raise RuntimeError(f"Upload failed: {e.code} {e.read().decode()}")

    # Delete
    delete_url = f"{url}/storage/v1/object/{bucket}/{storage_path}"
    req = Request(delete_url, method="DELETE")
    req.add_header("Authorization", f"Bearer {key}")
    try:
        resp = urlopen(req)
        print("[supabase] Delete OK:", resp.read().decode())
    except HTTPError as e:
        raise RuntimeError(f"Delete failed: {e.code} {e.read().decode()}")


def keep_alive_mongodb():
    from pymongo import MongoClient

    uri = os.environ["MONGODB_URI"]
    client = MongoClient(uri)
    db = client.get_database()
    collection = db["test"]

    result = collection.insert_one({
        "keepAlive": True,
        "message": "keep-alive ping",
    })
    print(f"[mongodb] Inserted doc {result.inserted_id}")

    collection.drop()
    print("[mongodb] Dropped 'test' collection")
    client.close()


if __name__ == "__main__":
    task = sys.argv[1] if len(sys.argv) > 1 else ""
    if task == "supabase":
        keep_alive_supabase()
    elif task == "mongodb":
        keep_alive_mongodb()
    else:
        print(f"Usage: python keepalive.py <supabase|mongodb>", file=sys.stderr)
        sys.exit(1)

File: scripts/keepalive.mjs
import { createClient } from "@supabase/supabase-js";
import mongoose from "mongoose";
import { readFileSync } from "fs";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const __dirname = dirname(fileURLToPath(import.meta.url));

const task = process.argv[2];

if (!task || !["supabase", "mongodb"].includes(task)) {
  console.error("Usage: node keepalive.js <supabase|mongodb>");
  process.exit(1);
}

// ── Supabase keep-alive ─────────────────────────────────────────────
async function keepAliveSupabase() {
  const url = process.env.SUPABASE_URL;
  const key = process.env.SUPABASE_SERVICE_ROLE_KEY;

  if (!url || !key) {
    throw new Error("SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY must be set");
  }

  const supabase = createClient(url, key);
  const bucket = "pdfs";
  const storagePath = "keepalive/sample.pdf";
  const pdfPath = join(__dirname, "sample.pdf");
  const pdfBuffer = readFileSync(pdfPath);

  console.log("[supabase] Uploading sample.pdf ...");
  const { error: uploadErr } = await supabase.storage
    .from(bucket)
    .upload(storagePath, pdfBuffer, {
      contentType: "application/pdf",
      upsert: true,
    });

  if (uploadErr) throw new Error(`Upload failed: ${uploadErr.message}`);
  console.log("[supabase] Upload OK");

  console.log("[supabase] Deleting sample.pdf ...");
  const { error: deleteErr } = await supabase.storage
    .from(bucket)
    .remove([storagePath]);

  if (deleteErr) throw new Error(`Delete failed: ${deleteErr.message}`);
  console.log("[supabase] Delete OK — done");
}

// ── MongoDB keep-alive ──────────────────────────────────────────────
async function keepAliveMongoDB() {
  const uri = process.env.MONGODB_URI;
  if (!uri) throw new Error("MONGODB_URI must be set");

  await mongoose.connect(uri);
  console.log("[mongodb] Connected");

  const db = mongoose.connection.db;
  const collection = db.collection("test");

  const doc = {
    keepAlive: true,
    timestamp: new Date(),
    message: "keep-alive ping",
  };

  const insertResult = await collection.insertOne(doc);
  console.log(`[mongodb] Inserted doc ${insertResult.insertedId}`);

  await collection.drop();
  console.log("[mongodb] Dropped 'test' collection — done");

  await mongoose.disconnect();
}

// ── Run ─────────────────────────────────────────────────────────────
const runners = { supabase: keepAliveSupabase, mongodb: keepAliveMongoDB };

runners[task]()
  .then(() => process.exit(0))
  .catch((err) => {
    console.error(`[keepalive] ${task} failed:`, err.message);
    process.exit(1);
  });
